File: sdks/javascript/src/mesh.ts
/**
 * 3D mesh and material components.
 */

import { Vec3 } from './math';

/**
 * 3D mesh component.
 */
export class Mesh {
  private meshType: string;

  private constructor(meshType: string) {
    this.meshType = meshType;
  }

  /**
   * Create a cube mesh.
   * 
   * @param size - Size of the cube
   * @returns A new cube mesh
   */
  static cube(size: number = 1.0): Mesh {
    const mesh = new Mesh('cube');
    (mesh as any).size = size;
    return mesh;
  }

  /**
   * Create a sphere mesh.
   * 
   * @param radius - Radius of the sphere
   * @param subdivisions - Number of subdivisions
   * @returns A new sphere mesh
   */
  static sphere(radius: number = 1.0, subdivisions: number = 32): Mesh {
    const mesh = new Mesh('sphere');
    (mesh as any).radius = radius;
    (mesh as any).subdivisions = subdivisions;
    return mesh;
  }

  /**
   * Create a plane mesh.
   * 
   * @param size - Size of the plane
   * @returns A new plane mesh
   */
  static plane(size: number = 10.0): Mesh {
    const mesh = new Mesh('plane');
    (mesh as any).size = size;
    return mesh;
  }

  toString(): string {
    return `Mesh(type='${this.meshType}')`;
  }
}

/**
 * Options for creating a material.
 */
export interface MaterialOptions {
  /** Albedo color */
  albedo?: [number, number, number];
  /** Metallic value */
  metallic?: number;
  /** Roughness value */
  roughness?: number;
}

/**
 * PBR material.
 */
export class Material {
  /** Albedo color */
  albedo: [number, number, number];
  
  /** Metallic value */
  metallic: number;
  
  /** Roughness value */
  roughness: number;

  /**
   * Create a new material.
   * 
   * @param options - Material options
   */
  constructor(options: MaterialOptions = {}) {
    this.albedo = options.albedo || [1.0, 1.0, 1.0];
    this.metallic = options.metallic ?? 0.5;
    this.roughness = options.roughness ?? 0.5;
  }

  /**
   * Create a standard PBR material.
   * 
   * @param options - Material options
   * @returns A new standard material
   */
  static standard(options: MaterialOptions = {}): Material {
    return new Material(options);
  }

  toString(): string {
    return `Material(albedo=${this.albedo}, metallic=${this.metallic}, roughness=${this.roughness})`;
  }
}

/**
 * Options for creating a 3D camera.
 */
export interface Camera3DOptions {
  /** Camera position */
  position?: Vec3;
  /** Look at target */
  lookAt?: Vec3;
  /** Field of view */
  fov?: number;
}

/**
 * 3D perspective camera.
 */
export class Camera3D {
  /** Camera position */
  position: Vec3;
  
  /** Look at target */
  lookAt: Vec3;
  
  /** Field of view */
  fov: number;

  /**
   * Create a new 3D camera.
   * 
   * @param options - Camera options
   */
  constructor(options: Camera3DOptions = {}) {
    this.position = options.position || new Vec3(0, 5, 10);
    this.lookAt = options.lookAt || Vec3.zero();
    this.fov = options.fov ?? 60.0;
  }

  toString(): string {
    return `Camera3D(pos=${this.position}, lookAt=${this.lookAt})`;
  }
}

/**
 * Options for creating a point light.
 */
export interface PointLightOptions {
  /** Light position */
  position?: Vec3;
  /** Light intensity */
  intensity?: number;
}

/**
 * Point light source.
 */
export class PointLight {
  /** Light position */
  position: Vec3;
  
  /** Light intensity */
  intensity: number;

  /**
   * Create a new point light.
   * 
   * @param options - Light options
   */
  constructor(options: PointLightOptions = {}) {
    this.position = options.position || Vec3.zero();
    this.intensity = options.intensity ?? 1000.0;
  }

  toString(): string {
    return `PointLight(pos=${this.position}, intensity=${this.intensity})`;
  }
}

/**
 * Options for creating a directional light.
 */
export interface DirectionalLightOptions {
  /** Light direction */
  direction?: Vec3;
  /** Light intensity */
  intensity?: number;
}

/**
 * Directional light source.
 */
export class DirectionalLight {
  /** Light direction */
  direction: Vec3;
  
  /** Light intensity */
  intensity: number;

  /**
   * Create a new directional light.
   * 
   * @param options - Light options
   */
  constructor(options: DirectionalLightOptions = {}) {
    this.direction = options.direction || new Vec3(0, -1, 0);
    this.intensity = options.intensity ?? 1.0;
  }

  toString(): string {
    return `DirectionalLight(dir=${this.direction}, intensity=${this.intensity})`;
  }
}

/**
 * Options for creating a spot light.
 */
export interface SpotLightOptions {
  /** Light position */
  position?: Vec3;
  /** Light direction */
  direction?: Vec3;
  /** Light intensity */
  intensity?: number;
  /** Cone angle */
  angle?: number;
}

/**
 * Spot light source.
 */
export class SpotLight {
  /** Light position */
  position: Vec3;
  
  /** Light direction */
  direction: Vec3;
  
  /** Light intensity */
  intensity: number;
  
  /** Cone angle */
  angle: number;

  /**
   * Create a new spot light.
   * 
   * @param options - Light options
   */
  constructor(options: SpotLightOptions = {}) {
    this.position = options.position || Vec3.zero();
    this.direction = options.direction || new Vec3(0, -1, 0);
    this.intensity = options.intensity ?? 1000.0;
    this.angle = options.angle ?? 45.0;
  }

  toString(): string {
    return `SpotLight(pos=${this.position}, intensity=${this.intensity})`;
  }
}

File: sdks/python/windjammer_sdk/mesh.py
"""
3D rendering for Windjammer.

This module provides 3D rendering components: meshes, materials, cameras, and lights.
"""

from typing import Optional
from .math import Vec3, Color
from .ffi import _lib, WjMesh, WjMaterial, WjCamera3D, WjPointLight, WjDirectionalLight, _check_error


class Mesh:
    """3D mesh component."""
    
    def __init__(self, mesh_type: str = "cube"):
        """
        Create a new mesh.
        
        Args:
            mesh_type: Type of mesh ("cube", "sphere", "plane", "custom")
        """
        self.mesh_type = mesh_type
        self._handle: Optional[WjMesh] = None
        # TODO: Create mesh with FFI
    
    @staticmethod
    def cube(size: float = 1.0) -> 'Mesh':
        """Create a cube mesh."""
        mesh = Mesh("cube")
        mesh.size = size
        return mesh
    
    @staticmethod
    def sphere(radius: float = 0.5, subdivisions: int = 32) -> 'Mesh':
        """Create a sphere mesh."""
        mesh = Mesh("sphere")
        mesh.radius = radius
        mesh.subdivisions = subdivisions
        return mesh
    
    @staticmethod
    def plane(size: float = 1.0) -> 'Mesh':
        """Create a plane mesh."""
        mesh = Mesh("plane")
        mesh.size = size
        return mesh
    
    def __repr__(self) -> str:
        return f"Mesh(type='{self.mesh_type}')"


class Material:
    """PBR material for 3D rendering."""
    
    def __init__(
        self,
        albedo: Optional[Color] = None,
        metallic: float = 0.0,
        roughness: float = 0.5,
        emissive: Optional[Color] = None
    ):
        """
        Create a new PBR material.
        
        Args:
            albedo: Base color
            metallic: Metallic factor (0.0 = dielectric, 1.0 = metal)
            roughness: Roughness factor (0.0 = smooth, 1.0 = rough)
            emissive: Emissive color
        """
        self.albedo = albedo or Color.white()
        self.metallic = metallic
        self.roughness = roughness
        self.emissive = emissive or Color(0, 0, 0, 1)
        self._handle: Optional[WjMaterial] = None
        # TODO: Create material with FFI
    
    def __repr__(self) -> str:
        return f"Material(albedo={self.albedo}, metallic={self.metallic}, roughness={self.roughness})"


class Camera3D:
    """3D perspective camera."""
    
    def __init__(
        self,
        position: Optional[Vec3] = None,
        look_at: Optional[Vec3] = None,
        fov: float = 60.0,
        near: float = 0.1,
        far: float = 1000.0
    ):
        """
        Create a new 3D camera.
        
        Args:
            position: Camera position in world space
            look_at: Point the camera is looking at
            fov: Field of view in degrees
            near: Near clipping plane distance
            far: Far clipping plane distance
        """
        self.position = position or Vec3(0, 0, 10)
        self.look_at = look_at or Vec3(0, 0, 0)
        self.fov = fov
        self.near = near
        self.far = far
        self._handle: Optional[WjCamera3D] = None
        # TODO: Create camera with FFI
    
    def __repr__(self) -> str:
        return f"Camera3D(pos={self.position}, look_at={self.look_at}, fov={self.fov})"


class PointLight:
    """Point light source (omnidirectional)."""
    
    def __init__(
        self,
        position: Optional[Vec3] = None,
        color: Optional[Color] = None,
        intensity: float = 1000.0,
        radius: float = 10.0
    ):
        """
        Create a new point light.
        
        Args:
            position: Light position in world space
            color: Light color
            intensity: Light intensity in lumens
            radius: Light radius (attenuation)
        """
        self.position = position or Vec3(0, 5, 0)
        self.color = color or Color.white()
        self.intensity = intensity
        self.radius = radius
        self._handle: Optional[WjPointLight] = None
        # TODO: Create light with FFI
    
    def __repr__(self) -> str:
        return f"PointLight(pos={self.position}, color={self.color}, intensity={self.intensity})"


class DirectionalLight:
    """Directional light source (like the sun)."""
    
    def __init__(
        self,
        direction: Optional[Vec3] = None,
        color: Optional[Color] = None,
        intensity: float = 1000.0
    ):
        """
        Create a new directional light.
        
        Args:
            direction: Light direction (normalized)
            color: Light color
            intensity: Light intensity in lux
        """
        self.direction = direction or Vec3(0, -1, 0)
        self.color = color or Color.white()
        self.intensity = intensity
        self._handle: Optional[WjDirectionalLight] = None
        # TODO: Create light with FFI
    
    def __repr__(self) -> str:
        return f"DirectionalLight(dir={self.direction}, color={self.color}, intensity={self.intensity})"


class PostProcessing:
    """Post-processing effects control."""
    
    @staticmethod
    def enable_hdr(enabled: bool = True):
        """Enable/disable HDR rendering."""
        # TODO: Use FFI
        pass
    
    @staticmethod
    def enable_bloom(enabled: bool = True):
        """Enable/disable bloom effect."""
        # TODO: Use FFI
        pass
    
    @staticmethod
    def enable_ssao(enabled: bool = True):
        """Enable/disable SSAO (Screen-Space Ambient Occlusion)."""
        # TODO: Use FFI
        pass
    
    @staticmethod
    def set_tone_mapping(mode: str = "ACES"):
        """
        Set tone mapping mode.
        
        Args:
            mode: Tone mapping mode ("None", "Reinhard", "ACES", "Filmic")
        """
        # TODO: Use FFI
        pass
    
    @staticmethod
    def set_color_grading(temperature: float = 0.0, saturation: float = 0.0, contrast: float = 0.0):
        """
        Set color grading parameters.
        
        Args:
            temperature: Color temperature adjustment (-1.0 to 1.0)
            saturation: Saturation adjustment (-1.0 to 1.0)
            contrast: Contrast adjustment (-1.0 to 1.0)
        """
        # TODO: Use FFI
        pass
